### PR TITLE
kubelet: Update engine version parsing to handle semantic versioning

### PR DESCRIPTION
**Pull request by krobertson**
This updates the dockertools.dockerVersion to use a semantic versioning
library to more gracefully support engine versions which include
additional version fields.

Previously, go-dockerclient's APIVersion struct was use which only
handles plain numeric x.y.z version strings. With #19675, the library
was now used on the Docker engine string, however it is possible for the
engine string to include include additional information for beta, rc, or
distro specific builds.

This PR also enables the TestDockerRuntimeVersion test which was
previously just a FIXME and updates it to pass, and be used to test the
version string that cause #20005.

This negates the need for fsouza/go-dockerclient#451, since even with
that change, if a user was running Docker 1.10.0-rc1, this would cause
the kubelet to report it as simply 1.10.0.

### CODE REVIEW

**Comment by krobertson** [comment]
cc @timothysc @dchen1107

**Comment by timothysc** [comment]
checking it out.

File: pkg/kubelet/kubelet_test.go [hunk 1]
```diff
@@ -2719,26 +2719,59 @@ func TestUpdateNewNodeOutOfDiskStatusWithTransitionFrequency(t *testing.T) {
...
 	fakeRuntime := testKubelet.fakeRuntime
 	fakeRuntime.RuntimeType = "docker"
-	fakeRuntime.VersionInfo = "1.5.0"
-	fakeRuntime.APIVersionInfo = "1.18"
+	fakeRuntime.VersionInfo = "1.10.0-rc1-fc24"
+	fakeRuntime.APIVersionInfo = "1.22"
 	kubeClient := testKubelet.fakeKubeClient
 	kubeClient.ReactionChain = testclient.NewSimpleFake(&api.NodeList{Items: []api.Node{
-		{ObjectMeta: api.ObjectMeta{Name: testKubeletHostname}},
+		{
```
Comment (timothysc) [inline pkg/kubelet/kubelet_test.go]: So some of these testing changes seem orthogonal to the original issue.  I would limit the scope only the version change.

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -944,18 +945,47 @@ func (dm *DockerManager) podInfraContainerChanged(pod *api.Pod, podInfraContaine
 	return podInfraContainerStatus.Hash != kubecontainer.HashContainer(expectedPodInfraContainer), nil
 }
 
-type dockerVersion docker.APIVersion
+// dockerVersion implementes kubecontainer.Version interface by implementing
+// Compare() and String() (which is implemented by the underlying semver.Version)
+// TODO: this code is the same as rktVersion and may make sense to be moved to
```
Comment (timothysc) [inline pkg/kubelet/dockertools/manager.go]: @dchen1107 | @vishh ?

File: pkg/kubelet/kubelet_test.go (hunk 1 above)
Comment (krobertson) [inline pkg/kubelet/kubelet_test.go]: I wanted to add a test case to cover this, and in lieu of adding a new testing exactly like this one, I re-enabled this test and updated it. Some of its mocks were outdated and incomplete, but didn't want to solve the original issue without ensuring it is covered in a test.

The mocks are updated from other tests I had to touch in the previous PR. So a bit copypasta, but is also consistent.

File: pkg/kubelet/dockertools/manager.go [hunk 3]
```diff
@@ -944,18 +945,47 @@ func (dm *DockerManager) podInfraContainerChanged(pod *api.Pod, podInfraContaine
...
+	if err != nil {
+		return dockerVersion{}, err
+	}
+	return dockerVersion{sem}, nil
+}
+
+func (r dockerVersion) Compare(other string) (int, error) {
+	v, err := semver.NewVersion(other)
+	if err != nil {
+		return -1, err
```
Comment (timothysc) [inline pkg/kubelet/dockertools/manager.go]: I don't know the semantics of Compare operation, but returning -1 on the error seems kind of odd to me, but meh.

**Comment by timothysc** [comment]
minor comments, and this change does work on our targets. 

```
$ kubectl describe nodes
...
 Container Runtime Version:	docker://1.10.0-rc1-fc24
```

File: pkg/kubelet/dockertools/manager.go [hunk 4]
```diff
@@ -31,6 +31,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/coreos/go-semver/semver"
```
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: @thockin: You favorite `semver` :)

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: A TODO is OK for now, although a subsequent refactoring PR would be much appreciated :)

**Comment by vishh** [comment]
LGTM. Kindly respond to @timothysc's comments and then I will add the LGTM label.

Thanks for the quick turnaround @krobertson!

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (krobertson) [inline pkg/kubelet/dockertools/manager.go]: I can. Wanted to call it out, but wanted this change in in short order.

Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: SGTM!

File: pkg/kubelet/dockertools/manager.go (hunk 3 above)
Comment (krobertson) [inline pkg/kubelet/dockertools/manager.go]: The `rktVersion` struct does the same: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/rkt/version.go#L45

Any preference on what to do? I'd go either way.

**Comment by vishh** [comment]
@krobertson: unit tests are failing. PTAL

**Comment by krobertson** [comment]
I haven't tried using the `test-cmd.sh` script before. I will take a look. From the build logs, I see:

```
E0122 23:43:43.506051   24660 manager.go:1016] docker: failed to parse docker server version "":  is not in dotted-tri format
```

For some reason, the engine string is blank. I can see elsewhere in the log that cAdvisor is showing it is Docker 1.9.1, which is what I'm using on my local cluster. I'll look some more.

**Comment by vishh** [comment]
FYI: This could be the integration tests where we stub docker daemon.

**Comment by krobertson** [comment]
Ah! Thanks for the tip. Saw the test stubs a `fake://` docker endpoint, so think I got it. Running locally to see.

**Comment by krobertson** [comment]
Should I put in a separate commit, or keep it squashed down to 1?

**Comment by vishh** [comment]
Separate commit SGTM

**Comment by krobertson** [comment]
Pushed up a 2nd commit which updates the `fake://` mock to include the engine version. I couldn't get `test-cmd.sh` to completely pass for me locally, since was failing on some other file check. I don't know if it was meant to be ran inside of the `run.sh` script.

The e2e tests posted that they failed too, but not sure where the error is. Saw ginko failed, but it also says:

```
Ran 119 of 215 Specs in 372.699 seconds
FAIL! -- 117 Passed | 0 Failed | 2 Pending | 94 Skipped 
```

**Comment by timothysc** [comment]
@vishh please label as P0, developers are still broken on fedora.

**Comment by vishh** [comment]
@timothysc: Done.

**Comment by timothysc** [comment]
many thx.
